### PR TITLE
experimental support for wrapping django middleware

### PR DESCRIPTION
**Pull request by piquadrat**

### CODE REVIEW

File: opbeat/base.py [hunk 1]
```diff
@@ -179,6 +179,10 @@ def __init__(self, organization_id=None, app_id=None, secret_token=None,
 
         self.processors = processors or defaults.PROCESSORS
         self.module_cache = ModuleProxyCache()
+        if wrap_django_middleware is not None:
+            self.wrap_django_middleware = wrap_django_middleware
+        else:
+            self.wrap_django_middleware = defaults.WRAP_DJANGO_MIDDLEWARE
```
Comment (roncohen) [inline opbeat/base.py]: I think this belongs in `contrib.django.client`

Comment (piquadrat) [inline opbeat/base.py]: True dat!

File: opbeat/contrib/django/middleware/__init__.py [hunk 2]
```diff
@@ -30,14 +37,22 @@ def _is_ignorable_404(uri):
 
 class Opbeat404CatchMiddleware(object):
     def process_response(self, request, response):
-        if response.status_code != 404 or _is_ignorable_404(request.get_full_path()):
+        if response.status_code != 404 or _is_ignorable_404(
+                request.get_full_path()
+        ):
```
Comment (roncohen) [inline opbeat/contrib/django/middleware/__init__.py]: I'd prefer:
```
if (response.status_code != 404 or
       _is_ignorable_404(request.get_full_path())):
    ...
```

File: opbeat/contrib/django/middleware/__init__.py [hunk 3]
```diff
@@ -45,50 +60,108 @@ def process_response(self, request, response):
         return response
 
 
-class OpbeatAPMMiddleware(object):
-    # Create a thread local variable to store the session in for logging
-    thread_local = threading.local()
+def get_name_from_middleware(wrapped, instance):
+    name = [type(instance).__name__, wrapped.__name__]
+    if type(instance).__module__:
+        name.insert(0, type(instance).__module__)
```
Comment (roncohen) [inline opbeat/contrib/django/middleware/__init__.py]: `name = type(instance).__module__ + name` ?

Comment (piquadrat) [inline opbeat/contrib/django/middleware/__init__.py]: `.__module__` might be `None`, hence the check

https://docs.python.org/2/reference/datamodel.html#index-36

File: tests/contrib/django/django_tests.py [hunk 4]
```diff
@@ -527,6 +529,77 @@ def test_request_metrics(self):
...
+                'django.middleware.common.CommonMiddleware',
+            ],
+            PREPEND_WWW=True,
+        ):
+            self.client.get(reverse('opbeat-no-error'))
+        timed_requests = self.opbeat._requests_store.get_all()
+        timing = timed_requests[0]
+        self.assertEqual(
+            timing['transaction'],
+            'django.middleware.common.CommonMiddleware.process_request'
```
Comment (roncohen) [inline tests/contrib/django/django_tests.py]: Any way we can give even more info? It seems like it would be useful to know when it was a PREPEND_WWW and when it was APPEND_SLASH?

File: tests/contrib/django/urls.py [hunk 5]
```diff
@@ -15,6 +15,8 @@ def handler500(request):
 
 urlpatterns = patterns('',
     url(r'^no-error$', 'tests.contrib.django.views.no_error', name='opbeat-no-error'),
+    url(r'^no-error-slash/$', 'tests.contrib.django.views.no_error', name='opbeat-no-error-slash'),
+    url(r'^login-required$', 'tests.contrib.django.views.login_required_view', name='opbeat-login-required'),
```
Comment (roncohen) [inline tests/contrib/django/urls.py]: I this used anywhere?

File: opbeat/contrib/django/middleware/__init__.py (hunk 3 above)
Comment (roncohen) [inline opbeat/contrib/django/middleware/__init__.py]: let me rephrase:
```
if type(instance).__module__:
    name = [type(instance).__module__] + name
```

File: tests/contrib/django/django_tests.py (hunk 4 above)
Comment (piquadrat) [inline tests/contrib/django/django_tests.py]: Only if we do special case handling for `CommonMiddleware`, and then infer from the redirect URL if it was `PREPEND_WWW` or `APPEND_SLASH` that caused the redirect. I'm not sure if that's worth it, tbh

File: opbeat/contrib/django/middleware/__init__.py (hunk 3 above)
Comment (piquadrat) [inline opbeat/contrib/django/middleware/__init__.py]: Ah, got it. Yep, that looks cleaner

**Comment by roncohen** [comment]
Mosltly minor stuff. Other than those :ship: :shipit:
